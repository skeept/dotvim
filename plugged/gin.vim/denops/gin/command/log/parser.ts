const defaultPattern =
  /^(?:[ *\|\\\/+\-=<>]*|[ *\|\\\/+\-=<>]*commit (?:[+\-=<>] )?)?([a-fA-F0-9]+)\b/;

export interface GitLogResult {
  entries: Entry[];
}

export type Entry = {
  commit: string;
};

export function parse(content: string[], pattern?: RegExp): GitLogResult {
  const entries: Entry[] = content.filter((v) => v).flatMap((record) => {
    const m = record.match(pattern ?? defaultPattern);
    if (m) {
      return [{
        commit: m[1],
      }];
    }
    return [];
  });
  return { entries };
}
