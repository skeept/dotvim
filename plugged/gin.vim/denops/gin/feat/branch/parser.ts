import { decodeUtf8 } from "../../util/text.ts";

const branchAliasPattern = /^\s{2}(remotes\/([^\/]+)\/(\S+))\s+-> (\S+)$/;
const remoteBranchPattern =
  /^\s{2}(remotes\/([^\/]+)\/(\S+))\s+([a-f0-9]+) (.*)$/;
const localBranchPattern =
  /^([* ]) (\(.*?\)|\S+)\s+([a-f0-9]+) (?:\[(\S+)(?:: [^\]]+)?\] )?(.*)$/;

export class GitBranchParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GitBranchParseError";
  }
}

type LocalBranch = {
  kind: "local";
  record: string;
  active: boolean;
  target: string;
  branch: string;
  commit: string;
  upstream?: string;
  message: string;
};

type RemoteBranch = {
  kind: "remote";
  record: string;
  target: string;
  remote: string;
  branch: string;
  commit: string;
  message: string;
};

type BranchAlias = {
  kind: "alias";
  record: string;
  target: string;
  remote: string;
  branch: string;
  origin: string;
};

export type Branch = LocalBranch | RemoteBranch | BranchAlias;

export interface GitBranchResult {
  branches: Branch[];
}

export function parse(bytes: Uint8Array): GitBranchResult {
  const text = decodeUtf8(bytes);
  const records: string[] = text.replace(/\n$/, "").split("\n");
  const branches: Branch[] = records.map((record) => {
    const m1 = record.match(branchAliasPattern);
    if (m1) {
      return {
        kind: "alias",
        record,
        target: m1[1],
        remote: m1[2],
        branch: m1[3],
        origin: m1[4],
      };
    }
    const m2 = record.match(remoteBranchPattern);
    if (m2) {
      return {
        kind: "remote",
        record,
        target: m2[1],
        remote: m2[2],
        branch: m2[3],
        commit: m2[4],
        message: m2[5],
      };
    }
    const m3 = record.match(localBranchPattern);
    if (m3) {
      return {
        kind: "local",
        record,
        active: m3[1] === "*",
        target: m3[2],
        branch: m3[2],
        commit: m3[3],
        upstream: m3[4] ?? undefined,
        message: m3[5],
      };
    }
    throw new Error(`Failed to parse 'git branch' record '${record}'`);
  });
  return { branches };
}
