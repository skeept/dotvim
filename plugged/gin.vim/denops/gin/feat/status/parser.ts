import { decodeUtf8, partition } from "../../util/text.ts";

export class GitStatusParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GitStatusParseError";
  }
}

export interface GitStatusResult {
  branch: BranchHeaders;
  entries: Entry[];
}

export function parse(
  bytes: Uint8Array,
): GitStatusResult {
  const it = partition(bytes);
  const entries: Entry[] = [];
  let branch: Partial<BranchHeaders> = {};
  while (true) {
    const ret = it.next();
    if (ret.done) {
      break;
    }
    const record = decodeUtf8(ret.value);
    if (!record) {
      continue;
    }
    switch (record.substring(0, 1)) {
      case "#": {
        branch = {
          ...branch,
          ...parseBranchHeaders(record),
        };
        break;
      }
      case "1": {
        entries.push(parseChangedEntry(record));
        break;
      }
      case "2": {
        const ret2 = it.next();
        if (ret2.done) {
          throw new GitStatusParseError(
            "No 'origPath' exists for renamed entry of 'Changed Tracked Entries'",
          );
        }
        const origPath = decodeUtf8(ret2.value);
        entries.push(parseRenamedEntry(record, origPath));
        break;
      }
      case "u": {
        entries.push(parseUnmergedEntry(record));
        break;
      }
      case "?": {
        entries.push(parseUntrackedEntry(record));
        break;
      }
      case "!": {
        entries.push(parseIgnoredEntry(record));
        break;
      }
      default: {
        throw new GitStatusParseError(
          `Unexpected record is detected: ${record}`,
        );
      }
    }
  }
  assertBranchHeaders(branch);
  return {
    branch,
    entries,
  };
}

export interface BranchHeaders {
  oid: string;
  head: string;
  upstream?: string;
  ahead?: number;
  behind?: number;
}

function assertBranchHeaders(
  branch: Partial<BranchHeaders>,
): asserts branch is BranchHeaders {
  if (!branch.oid || !branch.head) {
    throw new GitStatusParseError(`Not enough header`);
  }
}

export function parseBranchHeaders(
  record: string,
): Partial<BranchHeaders> {
  const columns = record.split(" ");
  if (columns[0] === "#" || columns.length >= 3) {
    const v = columns.slice(2).join(" ");
    switch (columns[1]) {
      case "branch.oid": {
        return {
          oid: v,
        };
      }
      case "branch.head": {
        return {
          head: v,
        };
      }
      case "branch.upstream": {
        return {
          upstream: v,
        };
      }
      case "branch.ab": {
        const [ahead, behind] = v.split(" ", 2);
        return {
          ahead: parseInt(ahead.substring(1), 10),
          behind: parseInt(behind.substring(1), 10),
        };
      }
    }
  }
  throw new GitStatusParseError(
    `Invalid 'Branch Headers' is specified: ${record}`,
  );
}

export interface ChangedEntry {
  kind: "changed";
  // The staged and unstaged XY values.
  XY: string;
  // The submodule state.
  sub: string;
  // The octal file mode in HEAD.
  mH: string;
  // The octal file mode in the index.
  mI: string;
  // The octal file mode in the worktree.
  mW: string;
  // The object name in HEAD.
  hH: string;
  // The object name in the index.
  hI: string;
  // The pathname.
  path: string;
}

export function parseChangedEntry(
  record: string,
): ChangedEntry {
  const columns = record.split(" ");
  if (columns[0] === "1" || columns.length >= 8) {
    return {
      kind: "changed",
      XY: columns[1],
      sub: columns[2],
      mH: columns[3],
      mI: columns[4],
      mW: columns[5],
      hH: columns[6],
      hI: columns[7],
      path: columns.slice(8).join(" "),
    };
  }
  throw new GitStatusParseError(
    `Invalid 'Changed Tracked Entries' is specified: ${record}`,
  );
}

export interface RenamedEntry {
  kind: "renamed";
  // The staged and unstaged XY values.
  XY: string;
  // The submodule state.
  sub: string;
  // The octal file mode in HEAD.
  mH: string;
  // The octal file mode in the index.
  mI: string;
  // The octal file mode in the worktree.
  mW: string;
  // The object name in HEAD.
  hH: string;
  // The object name in the index.
  hI: string;
  // The rename or copy score.
  Xscore: string;
  // The pathname.
  path: string;
  // The pathname in the commit at HEAD or in the index.
  origPath: string;
}

export function parseRenamedEntry(
  record: string,
  origPath: string,
): RenamedEntry {
  const columns = record.split(" ");
  if (columns[0] === "2" || columns.length >= 9) {
    return {
      kind: "renamed",
      XY: columns[1],
      sub: columns[2],
      mH: columns[3],
      mI: columns[4],
      mW: columns[5],
      hH: columns[6],
      hI: columns[7],
      Xscore: columns[8],
      path: columns.slice(9).join(" "),
      origPath,
    };
  }
  throw new GitStatusParseError(
    `Invalid 'Changed Tracked Entries' is specified: ${record}`,
  );
}

export interface UnmergedEntry {
  kind: "unmerged";
  // The conflict type.
  XY: string;
  // The submodule state.
  sub: string;
  // The octal file mode in stage 1.
  m1: string;
  // The octal file mode in stage 2.
  m2: string;
  // The octal file mode in stage 3.
  m3: string;
  // The octal file mode in the worktree.
  mW: string;
  // The object name in stage 1.
  h1: string;
  // The object name in stage 2.
  h2: string;
  // The object name in stage 3.
  h3: string;
  // The pathname.
  path: string;
}

export function parseUnmergedEntry(
  record: string,
): UnmergedEntry {
  const columns = record.split(" ");
  if (columns[0] === "u" || columns.length >= 10) {
    return {
      kind: "unmerged",
      XY: columns[1],
      sub: columns[2],
      m1: columns[3],
      m2: columns[4],
      m3: columns[5],
      mW: columns[6],
      h1: columns[7],
      h2: columns[8],
      h3: columns[9],
      path: columns.slice(10).join(" "),
    };
  }
  throw new GitStatusParseError(
    `Invalid 'Changed Tracked Entries' is specified: ${record}`,
  );
}

export interface UntrackedEntry {
  kind: "untracked";
  // The conflict type.
  XY: string;
  // The pathname.
  path: string;
}

export function parseUntrackedEntry(
  record: string,
): UntrackedEntry {
  const columns = record.split(" ");
  if (columns[0] === "?" || columns.length >= 2) {
    return {
      kind: "untracked",
      XY: "??",
      path: columns.slice(1).join(" "),
    };
  }
  throw new GitStatusParseError(
    `Invalid 'Other Items' is specified: ${record}`,
  );
}

export interface IgnoredEntry {
  kind: "ignored";
  // The conflict type.
  XY: string;
  // The pathname.
  path: string;
}

export function parseIgnoredEntry(
  record: string,
): IgnoredEntry {
  const columns = record.split(" ");
  if (columns[0] === "?" || columns.length >= 2) {
    return {
      kind: "ignored",
      XY: "!!",
      path: columns.slice(1).join(" "),
    };
  }
  throw new GitStatusParseError(
    `Invalid 'Other Items' is specified: ${record}`,
  );
}

export type Entry =
  | ChangedEntry
  | RenamedEntry
  | UnmergedEntry
  | UntrackedEntry
  | IgnoredEntry;
