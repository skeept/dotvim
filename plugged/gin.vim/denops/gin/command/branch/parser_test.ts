import { assertSnapshot } from "jsr:@std/testing@^1.0.0/snapshot";
import { parse } from "./parser.ts";

Deno.test("parse", async function (t): Promise<void> {
  await t.step("returns proper branches", async (t) => {
    const content = await Deno.readTextFile(
      new URL("./testdata/branch.txt", import.meta.url),
    );
    const result = parse(content.split("\n"));
    await assertSnapshot(t, result);
  });

  await t.step("returns proper branches even with worktree", async (t) => {
    const content = await Deno.readTextFile(
      new URL("./testdata/branch-with-worktree.txt", import.meta.url),
    );
    const result = parse(content.split("\n"));
    await assertSnapshot(t, result);
  });
});
