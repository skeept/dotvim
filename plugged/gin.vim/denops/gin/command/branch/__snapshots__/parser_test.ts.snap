export const snapshot = {};

snapshot[`parse > returns proper branches 1`] = `
{
  branches: [
    {
      active: false,
      branch: "gin-blame",
      commit: "b2a1c55",
      kind: "local",
      message: "WIP",
      record: "  gin-blame          b2a1c55 [main: ahead 1, behind 5] WIP",
      target: "gin-blame",
      upstream: "main",
      worktree: undefined,
    },
    {
      active: true,
      branch: "main",
      commit: "227f6c0",
      kind: "local",
      message: "Merge pull request #102 from lambdalisue/upgrade-git-browse",
      record: "* main               227f6c0 [origin/main] Merge pull request #102 from lambdalisue/upgrade-git-browse",
      target: "main",
      upstream: "origin/main",
      worktree: undefined,
    },
    {
      active: false,
      branch: "support-bang-bk",
      commit: "de1f346",
      kind: "local",
      message: ":+1: A bang (!) to forcibly open a buffer",
      record: "  support-bang-bk    de1f346 [support-bang: gone] :+1: A bang (!) to forcibly open a buffer",
      target: "support-bang-bk",
      upstream: "support-bang",
      worktree: undefined,
    },
    {
      active: false,
      branch: "support-read",
      commit: "ecf6a74",
      kind: "local",
      message: ":+1: Add \`postProcessor\` options",
      record: "  support-read       ecf6a74 [origin/support-read: gone] :+1: Add \`postProcessor\` options",
      target: "support-read",
      upstream: "origin/support-read",
      worktree: undefined,
    },
    {
      active: false,
      branch: "support-references",
      commit: "f1d2830",
      kind: "local",
      message: ":shower: Remove unused variables",
      record: "  support-references f1d2830 [origin/support-references: ahead 12, behind 8] :shower: Remove unused variables",
      target: "support-references",
      upstream: "origin/support-references",
      worktree: undefined,
    },
    {
      active: false,
      branch: "upgrade-deps",
      commit: "3ecec96",
      kind: "local",
      message: ":herb: Force a custom serializer of \`assertSnapshot\`",
      record: "  upgrade-deps       3ecec96 [origin/upgrade-deps: gone] :herb: Force a custom serializer of \`assertSnapshot\`",
      target: "upgrade-deps",
      upstream: "origin/upgrade-deps",
      worktree: undefined,
    },
    {
      active: false,
      branch: "upgrade-git-browse",
      commit: "3a534bc",
      kind: "local",
      message: ":package: Upgrade git-browse",
      record: "  upgrade-git-browse 3a534bc [origin/upgrade-git-browse: gone] :package: Upgrade git-browse",
      target: "upgrade-git-browse",
      upstream: "origin/upgrade-git-browse",
      worktree: undefined,
    },
  ],
}
`;

snapshot[`parse > returns proper branches even with worktree 1`] = `
{
  branches: [
    {
      active: false,
      branch: "fix-branch-on-worktree",
      commit: "3476d25",
      kind: "local",
      message: ":herb: Add branch parse test",
      record: "+ fix-branch-on-worktree 3476d25 (/Users/alisue/ghq/github.com/lambdalisue/my [work] (tree)) [main: ahead 1] :herb: Add branch parse test",
      target: "fix-branch-on-worktree",
      upstream: "main",
      worktree: "/Users/alisue/ghq/github.com/lambdalisue/my [work] (tree)",
    },
    {
      active: false,
      branch: "gin-blame",
      commit: "b2a1c55",
      kind: "local",
      message: "WIP",
      record: "  gin-blame              b2a1c55 [main: ahead 1, behind 5] WIP",
      target: "gin-blame",
      upstream: "main",
      worktree: undefined,
    },
    {
      active: true,
      branch: "main",
      commit: "227f6c0",
      kind: "local",
      message: "Merge pull request #102 from lambdalisue/upgrade-git-browse",
      record: "* main                   227f6c0 [origin/main] Merge pull request #102 from lambdalisue/upgrade-git-browse",
      target: "main",
      upstream: "origin/main",
      worktree: undefined,
    },
    {
      active: false,
      branch: "support-bang-bk",
      commit: "de1f346",
      kind: "local",
      message: ":+1: A bang (!) to forcibly open a buffer",
      record: "  support-bang-bk        de1f346 [support-bang: gone] :+1: A bang (!) to forcibly open a buffer",
      target: "support-bang-bk",
      upstream: "support-bang",
      worktree: undefined,
    },
    {
      active: false,
      branch: "support-read",
      commit: "ecf6a74",
      kind: "local",
      message: ":+1: Add \`postProcessor\` options",
      record: "  support-read           ecf6a74 [origin/support-read: gone] :+1: Add \`postProcessor\` options",
      target: "support-read",
      upstream: "origin/support-read",
      worktree: undefined,
    },
    {
      active: false,
      branch: "support-references",
      commit: "f1d2830",
      kind: "local",
      message: ":shower: Remove unused variables",
      record: "  support-references     f1d2830 [origin/support-references: ahead 12, behind 8] :shower: Remove unused variables",
      target: "support-references",
      upstream: "origin/support-references",
      worktree: undefined,
    },
    {
      active: false,
      branch: "upgrade-deps",
      commit: "3ecec96",
      kind: "local",
      message: ":herb: Force a custom serializer of \`assertSnapshot\`",
      record: "  upgrade-deps           3ecec96 [origin/upgrade-deps: gone] :herb: Force a custom serializer of \`assertSnapshot\`",
      target: "upgrade-deps",
      upstream: "origin/upgrade-deps",
      worktree: undefined,
    },
    {
      active: false,
      branch: "upgrade-git-browse",
      commit: "3a534bc",
      kind: "local",
      message: ":package: Upgrade git-browse",
      record: "  upgrade-git-browse     3a534bc [origin/upgrade-git-browse: gone] :package: Upgrade git-browse",
      target: "upgrade-git-browse",
      upstream: "origin/upgrade-git-browse",
      worktree: undefined,
    },
  ],
}
`;
